clang/lib/Sema/SemaConcept.cpp

```diff
@@ -972,8 +972,15 @@ static const Expr *SubstituteConstraintExpressionWithoutSatisfaction(
   // equivalence.
   LocalInstantiationScope ScopeForParameters(S);
   if (auto *FD = DeclInfo.getDecl()->getAsFunction())
-    for (auto *PVD : FD->parameters())
-      ScopeForParameters.InstantiatedLocal(PVD, PVD);
+    for (auto *PVD : FD->parameters()) {
+      if (!PVD->isParameterPack()) {
+        ScopeForParameters.InstantiatedLocal(PVD, PVD);
+        continue;
+      }
+      // Parameter packs should expand to a size-of-1 argument.
+      ScopeForParameters.MakeInstantiatedLocalArgPack(PVD);
+      ScopeForParameters.InstantiatedLocalPackArg(PVD, PVD);
+    }
```

Comment:
Also, this new comment should be clearer here.

We are trying to create a no-transformation scope, so in the end there is no actual pack expansion going on here: We are expanding an unexpanded pack into an expansion of one unexpanded pack, so we are back where we started and nothing was expanded. So this looks like a clever hack which avoided creating the SubstTemplateTypeParmPackType through a loophole.

Since concepts make this a valid use case, we could explore improving this abstraction, for example creating some sort of NonInstantiation scope where you just have to register which declarations are in scope, but shouldn't transform them.

Reply:
Actually, I think the better (future) approach here would be to ditch this scope and improve how FindInstantiatedDecl figures out which declarations should be part of the context being instantiated, and which are outside, so that it just naturally doesn't try to transform these declarations, just as for example it doesn't transform declarations at a template level above the context being instantiated.

Reply:
Thanks for the suggestion; I’m offline and will add more comments when I get to my computer.

IIRC, this future improvement could also benefit noexcept specifier comparison as it also requires an instantiation scope that just indicates, “We know this declaration is not instantiated; just go on and leave it an uninstantiated state”

Isolating the declaration instantiation from a TreeTransform seems to have inconvenienced other things. Maybe that is what was designed for templates in the first place, but it becomes cumbersome for cases like this and CWG2369, etc, and we need the on-demand instantiation.